aws-rds-dbinstance/src/main/java/software/amazon/rds/dbinstance/DeleteHandler.java

```diff
@@ -0,0 +1,73 @@
+import software.amazon.cloudformation.proxy.ProxyClient;
+import software.amazon.cloudformation.proxy.ResourceHandlerRequest;
+import software.amazon.cloudformation.resource.IdentifierUtils;
+
+public class DeleteHandler extends BaseHandlerStd {
+
+    private static final String SNAPSHOT_PREFIX = "snapshot-";
+    private static final int SNAPSHOT_MAX_LENGTH = 255;
+
+    protected ProgressEvent<ResourceModel, CallbackContext> handleRequest(
+            final AmazonWebServicesClientProxy proxy,
+            final ResourceHandlerRequest<ResourceModel> request,
+            final CallbackContext callbackContext,
+            final ProxyClient<RdsClient> proxyClient,
+            final Logger logger
+    ) {
+        return proxy.initiate("rds::delete-db-instance", proxyClient, request.getDesiredResourceState(), callbackContext)
+                .translateToServiceRequest(model -> {
+                    String finalDBSnapshotIdentifier = null;
+                    // https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-deletionpolicy.html
+                    // For AWS::RDS::DBInstance resources that don't specify the DBClusterIdentifier property, the default policy is Snapshot.
+                    if (BooleanUtils.isNotFalse(request.getSnapshotRequested())) {
+                        finalDBSnapshotIdentifier = model.getDBSnapshotIdentifier();
+                        if (StringUtils.isEmpty(finalDBSnapshotIdentifier)) {
+                            finalDBSnapshotIdentifier = IdentifierUtils.generateResourceIdentifier(
+                                    SNAPSHOT_PREFIX + Optional.ofNullable(request.getLogicalResourceIdentifier()).orElse("db-instance"),
+                                    Optional.ofNullable(request.getClientRequestToken()).orElse(""),
+                                    SNAPSHOT_MAX_LENGTH
+                            );
+                        }
+                    }
+                    return Translator.deleteDbInstanceRequest(model, finalDBSnapshotIdentifier);
+                })
+                .backoffDelay(BACK_OFF)
+                .makeServiceCall((deleteRequest, proxyInvocation) -> {
+                    if (callbackContext.isDeleting()) {
+                        return callbackContext.response("rds::delete-db-instance");
+                    }
+                    final DeleteDbInstanceResponse response = proxyInvocation.injectCredentialsAndInvokeV2(
+                            deleteRequest,
+                            proxyInvocation.client()::deleteDBInstance
+                    );
+                    callbackContext.setDeleting(true);
+                    return response;
+                })
+                .stabilize((deleteRequest, deleteResponse, proxyInvocation, model, context) -> {
+                    return isDbInstanceDeleted(proxyInvocation, model);
+                })
+                .handleError((deleteRequest, exception, client, model, context) -> {
```

Comment:
One of expected exceptions that deleteDBInstance can throw `DbSnapshotAlreadyExistsException` 
Should we handle it? I am not sure. Like
```
if (exception instanceof DbSnapshotAlreadyExistsException)
    return ProgressEvent.defaultFailureHandler(exception, HandlerErrorCode.AlreadyExists);
}
```

Reply:
This is a great question and it puzzled me for a while. I believe the intent here is to wrap RDS-specific 404 to CFN 404, which is expected to be handled by Uluru itself. The details of the failure would be available to the customer anyways. If I understand this part correctly, the intent is to distinguish retryable errors and dead-end ones. This must be taken through the prism of CFN peculiarity: a retryable from CLI/SDK perspective is unlike CFN retryable in a way: 4xx is not supposed to be retryable by the SDK because there is a conflicting state which the customer is supposed to resolve and then retry explicitly. 5xx is a retryable thing. In the CFN case the conflict is expecting a side effect from the customer, but the retry process is implicit unlike the previous one. This means that a 4xx error could expect the customer to eventually resolve the conflict and the automation will kick in. Except this 404: there is nothing to expect.
Having that said, I'm leaning towards handling 404 only in this context, hope this makes sense.